src/SignalR/clients/cpp/src/signalrclient/connection_impl.cpp

```diff
@@ -84,94 +84,139 @@ namespace signalr
             m_message_id = m_groups_token = m_connection_id = _XPLATSTR("");
         }
 
+        return start_negotiate(m_base_url, 0);
+    }
+
+    pplx::task<void> connection_impl::start_negotiate(const web::uri& url, int redirect_count)
+    {
+        if (redirect_count >= 100)
+        {
+            return pplx::task_from_exception<void>(signalr_exception(_XPLATSTR("too many redirects during negotiate")));
+        }
+
         pplx::task_completion_event<void> start_tce;
 
         auto weak_connection = weak_from_this();
 
         pplx::task_from_result()
-            .then([weak_connection]()
+            .then([weak_connection, url]()
+        {
+            auto connection = weak_connection.lock();
+            if (!connection)
             {
-                auto connection = weak_connection.lock();
-                if (!connection)
+                return pplx::task_from_exception<negotiation_response>(_XPLATSTR("connection no longer exists"));
+            }
+            return request_sender::negotiate(*connection->m_web_request_factory, url,
+                connection->m_query_string, connection->m_signalr_client_config);
+        }, m_disconnect_cts.get_token())
+            .then([weak_connection, start_tce, redirect_count, url](negotiation_response negotiation_response)
+        {
+            auto connection = weak_connection.lock();
+            if (!connection)
+            {
+                return pplx::task_from_exception<void>(_XPLATSTR("connection no longer exists"));
+            }
+
+            if (!negotiation_response.error.empty())
+            {
+                return pplx::task_from_exception<void>(signalr_exception(negotiation_response.error));
+            }
+
+            if (!negotiation_response.url.empty())
+            {
+                if (!negotiation_response.accessToken.empty())
                 {
-                    return pplx::task_from_exception<negotiation_response>(_XPLATSTR("connection no longer exists"));
+                    auto headers = connection->m_signalr_client_config.get_http_headers();
+                    headers.add(_XPLATSTR("Authorization"), _XPLATSTR("Bearer ") + negotiation_response.accessToken);
+                    connection->m_signalr_client_config.set_http_headers(headers);
                 }
-                return request_sender::negotiate(*connection->m_web_request_factory, connection->m_base_url,
-                    connection->m_query_string, connection->m_signalr_client_config);
-            }, m_disconnect_cts.get_token())
-            .then([weak_connection](negotiation_response negotiation_response)
+                return connection->start_negotiate(negotiation_response.url, redirect_count + 1);
```

Comment:
The redirect query string should replace m_query_string until the client is restarted. Either that, or remove the query string as something that can configured independently of the url like the other new clients.

Reply:
Removed the query string entirely 👍